Programs/python.c

```diff
@@ -55,6 +58,15 @@ main(int argc, char **argv)
     }
 
 #ifdef __ANDROID__
+    /* Locale coercion is not needed when the LANG or the LC_CTYPE environment
+     * variable is correctly set. */
+    env_loc = getenv("LANG");
+    if (env_loc == NULL || strncmp(env_loc, "C.UTF-8", 8) != 0) {
+        env_loc = getenv("LC_CTYPE");
+        if (env_loc == NULL || strncmp(env_loc, "C.UTF-8", 8) != 0) {
+            _Py_CoerceLegacyLocale();
+        }
+    }
```

Comment:
I was attempting to figure out how this differs from the existing logic in `_Py_LegacyLocaleDetected`, and it seems to me that this is requiring not only that the locale category `LC_ALL` be set to C.UTF-8 (which this code still forces unconditionally below), but also that either `LANG` or `LC_CTYPE` (or both) be set to C.UTF-8 in the environment.

If I've understood that correctly, then the only real effect of the call to `_Py_CoerceLegacyLocale` here that isn't already covered by the `setlocale` call below is to do `setenv("LC_CTYPE", "C.UTF-8", 1)`, just in a roundabout way that doesn't make it clear that that's all it is going to do.

If that's right, then I'd suggest just putting an unconditional `setenv("LC_ALL", "C.UTF-8", 1)` here to align with the `setlocale` call, and amende the comment to say something like "While setlocale doesn't check the environment variables, some other ported code may (e.g. in extension modules), so we make sure that the env vars match the forced locale configuration"

Reply:
> If I've understood that correctly, then the only real effect of the call to _Py_CoerceLegacyLocale here that isn't already covered by the setlocale call below is to do setenv("LC_CTYPE", "C.UTF-8", 1)

Actually another effect of _Py_CoerceLegacyLocale() is to print the _C_LOCALE_COERCION_WARNING (test_c_locale_coercion uses this warning to detect success or failure, hence the conditionals on the LANG and LC_CTYPE envt vars before calling _Py_CoerceLegacyLocale()).

The meaning of "locale coercion" may be understood as setting both the locale with setlocale() and the LC_CTYPE environment variable. My understanding is that this warning may be printed when warnings are enabled and Python is changing the LC_CTYPE environment variable.

With your proposed change, test_c_locale_coercion must be changed since _C_LOCALE_COERCION_WARNING is never printed on Andoid in that case. I do not have a strong opinion about this.

Reply:
> I do not have a strong opinion about this.

I mean that I do not have a strong opinion whether _C_LOCALE_COERCION_WARNING must be printed on Android or not.

Reply:
If I read the code correctly at [locale.cpp on bionic master branch (aka Android libc)](https://android.googlesource.com/platform/bionic/+/master/libc/bionic/locale.cpp#158) setlocale(LC_CTYPE, "") returns "C.UTF-8". But on Android API 24 it does return "C" as expected by the implementation of PEP 538.

Another point is that call_readline() in Modules/readline.c does call setlocale(LC_CTYPE, "") (and the reason I gave in [msg305848](https://bugs.python.org/msg305848) to explain why test_nonasci succeeds seems to be wrong).

For these reasons, I am starting to think we could use a new _Py_setlocale_from_env() function to replace all calls to setlocale(category, "") with a custom implementation for Android and returning setlocale(category, "") for all other platforms.

Reply:
Yeah, I think so - that way we can isolate the idiosyncrasies of different libc implementations a bit better.

As far as the locale coercion warning goes, my understanding is that Android really is UTF-8 everywhere, so we're not at significant risk of corrupting improperly encoded data and hence shouldn't print the warning. Instead, we're just adjusting the environment so that extension modules and CPython code that are assuming glibc-style locale handling see what they expect to see.

Reply:
However, making sure I understand the suggestion correctly:

* on most platforms, `_Py_setlocale_from_env(category)` would just call `setlocale(category, "")`
* on Android, we'd special case it such that if none of LC_ALL, LC_CATEGORY, or LANG were set, we'd pass "C.UTF-8" explicitly (otherwise we'd pass whatever the highest priority environment variable was set to).